src/realm/replication.cpp

```diff
@@ -62,36 +64,323 @@ Replication::version_type Replication::prepare_commit(version_type orig_version)
                                              Table::Type table_type)
 {
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Add %4 class '%1' with primary key property '%2' of %3",
+                    Group::table_name_to_class_name(name), pk_name, pk_type, table_type);
+    }
     REALM_ASSERT(table_type != Table::Type::Embedded);
     unselect_all();
     m_encoder.insert_group_level_table(tk); // Throws
 }
 
+void Replication::erase_class(TableKey tk, StringData table_name, size_t)
+{
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Remove class '%1'", Group::table_name_to_class_name(table_name));
+    }
+    unselect_all();
+    m_encoder.erase_class(tk); // Throws
+}
+
+
+void Replication::insert_column(const Table* t, ColKey col_key, DataType type, StringData col_name,
+                                Table* target_table)
+{
+    if (auto logger = get_logger()) {
+        const char* collection_type = "";
+        if (col_key.is_collection()) {
+            if (col_key.is_list()) {
+                collection_type = "list ";
+            }
+            else if (col_key.is_dictionary()) {
+                collection_type = "dictionary ";
+            }
+            else {
+                collection_type = "set ";
+            }
+        }
+        if (target_table) {
+            logger->log(util::Logger::Level::debug, "On class '%1': Add property '%2' %3linking '%4'",
+                        t->get_class_name(), col_name, collection_type, target_table->get_class_name());
+        }
+        else {
+            logger->log(util::Logger::Level::debug, "On class '%1': Add property '%2' %3of %4", t->get_class_name(),
+                        col_name, collection_type, type);
+        }
+    }
+    select_table(t);                  // Throws
+    m_encoder.insert_column(col_key); // Throws
+}
+
+void Replication::erase_column(const Table* t, ColKey col_key)
+{
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "On class '%1': Remove property '%2'", t->get_class_name(),
+                    t->get_column_name(col_key));
+    }
+    select_table(t);                 // Throws
+    m_encoder.erase_column(col_key); // Throws
+}
+
 void Replication::create_object(const Table* t, GlobalKey id)
 {
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Create object '%1'", t->get_class_name());
+    }
     select_table(t);                              // Throws
     m_encoder.create_object(id.get_local_key(0)); // Throws
 }
 
-void Replication::create_object_with_primary_key(const Table* t, ObjKey key, Mixed)
+void Replication::create_object_with_primary_key(const Table* t, ObjKey key, Mixed pk)
 {
+    if (auto logger = get_logger()) {
+        logger->log(util::Logger::Level::debug, "Create object '%1' with primary key %2", t->get_class_name(), pk);
```

Comment:
it's a bit inconsistent. Why do you omit key but log it still in other places below?

Reply:
I will try to be consistent to log key when we don't have primary key or path.